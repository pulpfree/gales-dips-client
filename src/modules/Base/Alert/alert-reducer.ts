/**
 * Alert Reducer
 *
 * Got most of this from link below
 * @see https://kentcdodds.com/blog/how-to-test-custom-react-hooks?ck_subscriber_id=635632627
 */

import React from 'react'

import { nanoid } from 'nanoid'

const CLEAR = 'CLEAR'
const DISMISS = 'DISMISS'
const SEND = 'SEND'

const DefaultTimeout = 4000

export type Color = 'success' | 'info' | 'warning' | 'error'

export type AlertItem = {
  id?: string
  setTimeout?: boolean
  message: string
  severity?: Color
  timeout?: number
  title?: string
}

export type AlertState = AlertItem[]

export type Action =
  | {
      type: 'SEND'
      payload: AlertItem
    }
  | {
      type: 'DISMISS'
      payload: string
    }
  | {
      type: 'CLEAR'
    }

export type AlertT = {
  alerts: AlertItem[]
  clear: () => void
  dismiss: (id: string) => void
  send: (alert: AlertItem) => void
}

const alertReducer = (state: AlertState, action: Action): AlertState => {
  switch (action.type) {
    case SEND: {
      const { id, setTimeout } = action.payload
      if (!id) {
        action.payload.id = nanoid()
      }
      action.payload.setTimeout = typeof setTimeout === 'undefined' ? true : false

      return [...state, action.payload]
    }

    case DISMISS: {
      if (!action.payload) {
        throw new Error('Missing alert id')
      }
      return state.filter((alert) => alert.id !== action.payload)
    }

    case CLEAR:
      return []

    default:
      throw new Error('Unhandled action type')
  }
}

export const initialState = {
  alerts: [],
}

export const useAlert = (): AlertT => {
  const [state, dispatch] = React.useReducer(alertReducer, [])

  const send = React.useCallback((alert: AlertItem) => {
    const timeout = alert.timeout || DefaultTimeout
    dispatch({ type: SEND, payload: alert })
    if (alert.setTimeout) {
      setTimeout(() => dispatch({ type: DISMISS, payload: alert.id! }), timeout) // eslint-disable-line @typescript-eslint/no-non-null-assertion
    }
  }, [])

  const dismiss = React.useCallback((id: string) => dispatch({ type: DISMISS, payload: id }), [])

  const clear = React.useCallback(() => dispatch({ type: CLEAR }), [])

  return { clear, dismiss, send, alerts: [...state] }
}
