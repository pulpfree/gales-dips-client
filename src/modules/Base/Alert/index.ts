export * from './AlertContext'
export * from './alert-reducer'
export * from './Alerts'
