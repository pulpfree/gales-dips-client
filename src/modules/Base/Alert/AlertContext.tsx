/**
 * Testing example @see: https://codesandbox.io/s/modern-sea-tx3ro?from-embed=&file=/src/components/Greeting.test.tsx
 */
import React from 'react'

import { AlertItem, useAlert } from './alert-reducer'

type State = AlertItem[]
type Dispatch = {
  alerts: AlertItem[]
  clear: () => void
  dismiss: (id: string) => void
  send: (error: AlertItem) => void
}

type AlertProviderProps = {
  children: React.ReactNode
}

const AlertStateContext = React.createContext<State | undefined>(undefined)
const AlertDispatchContext = React.createContext<Dispatch | undefined>(undefined)

export const AlertProvider = ({ children }: AlertProviderProps): JSX.Element => {
  const userHandler: Dispatch = useAlert()

  return (
    <AlertStateContext.Provider value={userHandler.alerts}>
      <AlertDispatchContext.Provider value={userHandler}>{children}</AlertDispatchContext.Provider>
    </AlertStateContext.Provider>
  )
}

export const useAlertState = (): State => {
  const context = React.useContext(AlertStateContext)
  if (context === undefined) {
    throw new Error('useAlertState must be used within a CountProvider')
  }
  return context
}

export const useAlertDispatch = (): Dispatch => {
  const context = React.useContext(AlertDispatchContext)
  if (context === undefined) {
    throw new Error('useAlertDispatch must be used within a CountProvider')
  }
  return context
}
