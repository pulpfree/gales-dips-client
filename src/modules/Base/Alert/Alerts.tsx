import React from 'react'

import { IconButton } from '@material-ui/core'
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles'
import CloseIcon from '@material-ui/icons/Close'
import MuiAlert, { AlertProps } from '@material-ui/lab/Alert'
import { AlertTitle } from '@material-ui/lab'

import { useAlertDispatch, useAlertState } from './AlertContext'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      alignItems: 'center',
      display: 'flex',
      flexDirection: 'column',
      margin: '0px 100px 0 100px',
      position: 'absolute',
      top: 65,
      width: 'calc(100vw - 200px)',
      // backgroundColor: 'gray',
      zIndex: 100,
      '& > * + *': {
        marginTop: theme.spacing(1),
      },
    },
    actionItem: {
      width: 600,
    },
  }),
)

export const Alerts = (): JSX.Element | null => {
  const context = useAlertState()
  const classes = useStyles()

  const Alert = (props: AlertProps) => {
    const dispatch = useAlertDispatch()

    return (
      <MuiAlert
        className={classes.actionItem}
        elevation={6}
        // severity={props.severity}
        variant='filled'
        action={
          <IconButton
            aria-label='close'
            color='inherit'
            size='small'
            onClick={() => dispatch.dismiss(props.id!)} // eslint-disable-line @typescript-eslint/no-non-null-assertion
          >
            <CloseIcon fontSize='inherit' />
          </IconButton>
        }
        {...props}
      />
    )
  }

  if (!context.length) return null

  return (
    <div className={classes.root}>
      {context.map((c) => (
        <Alert key={c.id} id={c.id} severity={c.severity}>
          {c.title && <AlertTitle>{c.title}</AlertTitle>}
          {c.message}
        </Alert>
      ))}
    </div>
  )
}
