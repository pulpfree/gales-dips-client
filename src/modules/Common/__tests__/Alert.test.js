/* eslint no-undef: "off" */

import React from 'react'
import { shallow } from 'enzyme'

import Alert from '../Alert'

describe('Alert', () => {

  let mountedC
  let props
  const reportC = () => {
    if (!mountedC) {
      mountedC = shallow(
        <Alert {...props} />
      )
    }
    return mountedC
  }

  beforeEach(() => {
    props = {
      authState: 'forgotPassword',
      onStateChange: () => {},
      children: '',
      type: '',
    }
    mountedC = undefined
  })

  it('matches snapshot', () => {
    expect(reportC()).toMatchSnapshot()
  })

})