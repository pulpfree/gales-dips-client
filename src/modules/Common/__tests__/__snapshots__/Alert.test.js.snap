// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Alert)
    authState="forgotPassword"
    onStateChange={[Function]}
    type=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "authState": "forgotPassword",
      "children": "",
      "classes": Object {
        "base": "Alert-base-1",
        "danger": "Alert-danger-2",
        "info": "Alert-info-4",
        "success": "Alert-success-3",
        "warning": "Alert-warning-5",
      },
      "onStateChange": [Function],
      "type": "",
    },
    "ref": null,
    "rendered": "",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "authState": "forgotPassword",
        "children": "",
        "classes": Object {
          "base": "Alert-base-1",
          "danger": "Alert-danger-2",
          "info": "Alert-info-4",
          "success": "Alert-success-3",
          "warning": "Alert-warning-5",
        },
        "onStateChange": [Function],
        "type": "",
      },
      "ref": null,
      "rendered": "",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
